# Assignment 2: Entity Resolution (Part 2)


## Objective

In Assignment 2 (Part 2), you will learn how to use Active Learning to address the entity resolution problem. After completing this assignment, you should be able to answer the following questions:

1. Why Active Learning?
2. How to implement uncertain sampling, a popular query strategy for Active Learning?
3. How to solve an ER problem using Active Learning?


## Active Learning

[Active learning](http://tiny.cc/al-wiki) is a certain type of ML algorithms that can train a high-quality ML model with small data-labeling cost. Its basic idea is quite easy to understand. Consider a typical supervised ML problem, which requires a (relatively) large training dataset. In the training dataset, there may be only a small number of data points that are beneficial to the trained ML model. In other words, labeling a small number of data points is enough to train a high-quality ML model. The goal of active learning is to help us to identify those data points.


In this assignment, we will develop an Active Learning approach for Entity Resolution. The following figure shows the architecture of an entity resolution solution. It consists of four major steps. **I will provide you the source code for Steps 1, 2, 4. Your job is to implement Step 3.**  

<img src="/content/img/arch.png", width=800/>

![title](/content/img/arch.png)

### Step 1. Read Data

Suppose we get a restaurant dataset `restaurant.csv`. The data has many duplicate restaurants.  For example, the first two rows shown below are duplicated (i.e., refer to the same real-world entity). You can check out all duplicate (matching) record pairs from `true_matches.json`.

In [1]:
import pandas as pd

df = pd.read_csv('restaurant.csv')
data = df.values.tolist()
print("(#Rows, #Cols) :", df.shape)
df.head(6)

(#Rows, #Cols) : (858, 5)


id                       name                  address         city  \
0   1  arnie morton's of chicago   435 s. la cienega blv.  los angeles   
1   2  arnie morton's of chicago  435 s. la cienega blvd.  los angeles   
2   3         art's delicatessen      12224 ventura blvd.  studio city   
3   4                 art's deli      12224 ventura blvd.  studio city   
4   5              hotel bel-air     701 stone canyon rd.      bel air   
5   6              bel-air hotel     701 stone canyon rd.      bel air   

          type  
0     american  
1  steakhouses  
2     american  
3        delis  
4  californian  
5  californian

In [2]:
truth = pd.read_json('true_matches.json')
print("(#Rows, #Cols) :", truth.shape)
truth.head(5)
truth[truth[0] == 5] # column values are id's of matched pairs. Each row indicates a pair.

(#Rows, #Cols) : (106, 2)


0  1
71  5  6

### Step 2. Similar Pairs

We first use a similarity-join algorithm to generate similar pairs.

Below is the code. After running the code, we get 678 similar pairs ordered by their similarity decreasingly.

In [3]:
from a2_utils import *

data = df.values.tolist()
simpairs = simjoin(data)

print("Num of Pairs: ", len(data)*(len(data)-1)/2)
print("Num of Similar Pairs: ", len(simpairs))
print("The Most Similar Pair: ", simpairs[0])

Num of Pairs:  367653.0
Num of Similar Pairs:  678
The Most Similar Pair:  ([170, "mary mac's tea room", '224 ponce de leon ave.', 'atlanta', 'southern/soul'], [169, "mary mac's tea room", '224 ponce de leon ave.', 'atlanta', 'southern'])


We can see that `simjoin` helps us remove the number of pairs from 367653 to 678. But, there are still many non-matching pairs in `simpairs` (see below).

In [4]:
print(simpairs[-1])

([764, "buzio's in the rio", '3700 w. flamingo rd.', 'las vegas', 'seafood'], [542, 'carnival world', '3700 w. flamingo rd.', 'las vegas', 'buffets'])


Next, we will use active learning to train a classifier, and then use the classifier to classify each pair in `simpairs` as either "matching" or "nonmatching".

### Step 3. Active Learning

Given a set of similar pairs, what you need to do next is to iteratively train a classifier to decide which pairs are truly matching. We are going to use [logistic regression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html) as our classifier.

#### Initialization

At the beginning, all the pairs are unlabeled. To initialize a model, we first pick up ten pairs and then label each pair using  the `crowdsourcing()` function. You can assume that `crowdsourcing()` will ask a crowd worker (e.g., on Amazon Mechanical Turk) to label a pair.


`crowdsourcing(pair)` is a function that simulates the use of crowdsourcing to label a pair
  
  - **Input:**	pair – A pair of records

  - **Output:**	Boolean –  *True*: The pair of records are matching; *False*: The pair of records are NOT matching;

Please use the following code to do the initialization.

In [5]:
from a2_utils import crowdsourcing

# choose the most/least similar five pairs as initial training data
init_pairs = simpairs[:5] + simpairs[-5:]
matches = []
nonmatches = []
for pair in init_pairs:
    is_match = crowdsourcing(pair)
    if is_match == True:
        matches.append(pair)
    else:
        nonmatches.append(pair)

print("Number of matches: ", len(matches))
print("Number of nonmatches: ", len(nonmatches))

Are they matching?
[170, "mary mac's tea room", '224 ponce de leon ave.', 'atlanta', 'southern/soul']
[169, "mary mac's tea room", '224 ponce de leon ave.', 'atlanta', 'southern']
Answer: Yes
Are they matching?
[88, 'manhattan ocean club', '57 w. 58th st.', 'new york city', 'seafood']
[87, 'manhattan ocean club', '57 w. 58th st.', 'new york', 'seafood']
Answer: Yes
Are they matching?
[112, 'san domenico', '240 central park s.', 'new york city', 'italian']
[111, 'san domenico', '240 central park s', 'new york', 'italian']
Answer: Yes
Are they matching?
[197, 'fleur de lys', '777 sutter st.', 'san francisco', 'french (new)']
[196, 'fleur de lys', '777 sutter st.', 'san francisco', 'french']
Answer: Yes
Are they matching?
[8, 'cafe bizou', '14016 ventura blvd.', 'sherman oaks', 'french bistro']
[7, 'cafe bizou', '14016 ventura blvd.', 'sherman oaks', 'french']
Answer: Yes
Are they matching?
[709, 'arcadia', '21 e. 62nd st.', 'new york city', 'american (new)']
[66, 'four seasons', '99 e. 5

Here is the only code you need to write in this assignment.


In [21]:
import pandas as pd
from sklearn.linear_model import LogisticRegression
from a2_utils import featurize, crowdsourcing


labeled_pairs = matches + nonmatches
unlabeled_pairs = [p for p in simpairs if p not in labeled_pairs]
iter_num = 5

#<-- Write Your Code -->

# Prepare X and y
X_labeled = pd.DataFrame([featurize(pair) for pair in labeled_pairs])
ones = [1 for x in range(5)]
zeros = [0 for x in range(5)]
y_labeled = ones + zeros

# train an initial model on labeled_pairs

model = LogisticRegression()
fitted_model = model.fit(X_labeled, y_labeled)

for i in range(iter_num):
    # Make predictions on the unlabeled data
    X_unlabeled = pd.DataFrame([featurize(pair) for pair in unlabeled_pairs])
    predictions = fitted_model.predict_proba(X_unlabeled)

    # Find the predictions closest to [0.5, 0.5]
    # Approach: Subtract 0.5 from the first element of every row,
    # then take the absolute val. The element closest to 0 will be the most uncertain
    # Why? Since an original value of 0.5 - 0.5 = 0.
    uncertainties = abs(predictions[:, 0] - 0.5)
    most_uncertain_index = uncertainties.argmin()

    # Crowdsource the label for the most uncertain pair
    most_uncertain_pair = unlabeled_pairs[most_uncertain_index]
    is_match = crowdsourcing(most_uncertain_pair)
    if is_match:
        y_labeled.append(1)
    else:
        y_labeled.append(0)

    # Update our list of pairs
    labeled_pairs.append(most_uncertain_pair)
    del unlabeled_pairs[most_uncertain_index]

    # Recreate the training data
    X_labeled = pd.DataFrame([featurize(pair) for pair in labeled_pairs])

    # Retrain the model
    fitted_model = model.fit(X_labeled, y_labeled)

Are they matching?
[817, 'ritz-carlton cafe (atlanta)', '181 peachtree st.', 'atlanta', 'american (new)']
[180, 'ritz-carlton restaurant', '181 peachtree st.', 'atlanta', 'french (classic)']
Answer: No
Are they matching?
[140, 'second street grill', '200 e. fremont st.', 'las vegas', 'pacific rim']
[139, 'second street grille', '200 e. fremont st.', 'las vegas', 'seafood']
Answer: Yes
Are they matching?
[817, 'ritz-carlton cafe (atlanta)', '181 peachtree st.', 'atlanta', 'american (new)']
[179, 'restaurant  ritz-carlton  atlanta', '181 peachtree st.', 'atlanta', 'continental']
Answer: No
Are they matching?
[177, 'dining room  ritz-carlton  buckhead', '3434 peachtree rd.', 'atlanta', 'international']
[175, 'cafe  ritz-carlton  buckhead', '3434 peachtree rd.', 'atlanta', 'ext 6108 international']
Answer: No
Are they matching?
[176, 'ritz-carlton cafe (buckhead)', '3434 peachtree rd. ne', 'atlanta', 'american (new)']
[175, 'cafe  ritz-carlton  buckhead', '3434 peachtree rd.', 'atlanta', '

**[Algorithm Description].**   Active learning has many [query strategies](http://tiny.cc/al-wiki-qs) to decide which data point should be labeled. You need to implement uncertain sampling. The algorithm trains an initial model on `labeled_pairs`. Then, it iteratively trains a model. At each iteration, it first applies the model to `unlabeled_pairs`, and makes a prediction on each unlabeled pair along with a probability, where the probability indicates the confidence of the prediction. After that, it selects the most uncertain pair (If there is still a tie, break it randomly),  and call the `crowdsroucing()` function to label the pair. After the pair is labeled, it updates `labeled_pairs` and `unlabeled_pairs`, and then retrain the model on `labeled_pairs`.

**[Input].**
- `labeled_pairs`: 10 labeled pairs (by default)
- `unlabeled_pairs`: 668 unlabeled pairs (by default)
- `iter_num`: 5 (by default)

**[Output].**
- `model`: A logistic regression model built by scikit-learn


### Step 4. Model Evaluation

After training an model, you can use the following code to evalute it.

In [22]:
import json
import numpy as np
from a2_utils import evaluate


sp_features = np.array([featurize(sp) for sp in simpairs])
label = model.predict(sp_features)
pair_label = zip(simpairs, label)

identified_matches = []
for pair, label in pair_label:
    if label == 1:
        identified_matches.append(pair)

precision, recall, fscore = evaluate(identified_matches)

print("Precision:", precision)
print("Recall:", recall)
print("Fscore:", fscore)


Precision: 1.0
Recall: 0.7452830188679245
Fscore: 0.854054054054054


## Submission

Complete the code in A2-2.ipynb, and submit it to the CourSys activity Assignment 2.